[SeanFoongjt] iP

SeanFoongjt opened the pull request:
# Elster
> "Plan for what it is difficult while it is easy, do what is great while it is small.” - Sun Tzu [(source)](https://www.goodreads.com/quotes/964849-plan-for-what-it-is-difficult-while-it-is-easy)

Elster is a chatbot that does it's best to help you out. It's
- Speedy
- Intuitive
- And ~~maybe~~ just a bit sarcastic

To experience this, all you have to do is to
1. download [it](https://github.com/SeanFoongjt/ip/tree/master)
2. run the program
3. add some tasks
4. ???
5. _Profit!_ 💰 

All for the price of __0 Dollarinos!__

Features:
- [x] Adding tasks
- [x] Removing tasks
- [ ] Reminders (sometime soon)
- [ ] Calendars (sometime soon)

If you're learning Java, you can use it to practice Java too! Here's the `main` method:
```java
public static void main(String[] args) {
        Path dataDir = Paths.get("data");
        new Elster(dataDir).run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SeanFoongjt committed:
Change name of Duke to Elster and add a goodbye

SeanFoongjt committed:
Add the echo function

SeanFoongjt committed:
Add list and add to list functions

SeanFoongjt committed:
Add mark and unmark function and implement classes

The main program is handling all functionalities. This includes
related behaviors.

The related behaviors can be encapsulated in a class which also reduces
code clutter

We should group the related functions of the task list and the task into
it's own classes.

This allows for better modularity of the code for future purposes and
also reduces code clutter.

SeanFoongjt committed:
Add deadline, event and todo tasks and inheritance

Task classes (e.g. DeadlineTask, EventTask etc.) have some common
behaviors with each other.

The common behaviors are causing duplication of code.

We can extract the similar behaviors into a superclass Task to support
polymorphism as well as reduce code duplication.

Let us create a superclass Task that the other tasks like deadlines and
events inherit from.

SeanFoongjt committed:
Add text-ui automated testing

SeanFoongjt committed:
Handle most common errors encountered

SeanFoongjt committed:
Add delete function

SeanFoongjt committed:
Add ability to read and write local files

SeanFoongjt committed:
Add write to file function for Todo Task

SeanFoongjt committed:
Add write file functions for deadline and event

SeanFoongjt committed:
Merge branch 'branch-Level-7'

SeanFoongjt committed:
Let deadline track deadline as LocalDateTime

SeanFoongjt committed:
Let event track start and end as LocalDateTime

SeanFoongjt committed:
Merge branch 'branch-Level-8'

SeanFoongjt committed:
Reformat to have Ui class

SeanFoongjt committed:
Add Elseptions (errors) and Storage

SeanFoongjt committed:
Update Ui, Storage and TaskList

SeanFoongjt committed:
Updated Ui, Elseption, Storage and TaskList

SeanFoongjt committed:
Fixed storage and added more OOP

SeanFoongjt committed:
Merge branch 'A-MoreOOP'

SeanFoongjt committed:
Add packages

SeanFoongjt committed:
Merge branch 'A-Packages'

SeanFoongjt committed:
Merge remote-tracking branch 'origin/add-gradle-support'

SeanFoongjt committed:
Add gradle build and automation

SeanFoongjt committed:
Add JUnit dependency and TaskListTest

SeanFoongjt committed:
Add tests for Deadline Task

SeanFoongjt committed:
Create a jar file for Elster

SeanFoongjt committed:
Add JavaDoc header comments for Task classes

SeanFoongjt committed:
Correct files to follow Java coding standard

SeanFoongjt committed:
Add findByDescription and descContain functions

SeanFoongjt committed:
Implement find by description in UI and Elster

SeanFoongjt committed:
Add Javadocs for Ui

SeanFoongjt committed:
Add Javadoc comments to TaskList

SeanFoongjt committed:
Merge branch 'A-CodingStandard'

SeanFoongjt committed:
Merge branch 'Level-9'

GuanpengR commented on src/main/java/elster/Elster.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    public void run() {
+        boolean byeSentinel = true;
```
Perhaps renaming the boolean?

GuanpengR commented on src/main/java/elster/Storage.java:
```diff
@@ -0,0 +1,102 @@
+     * @param taskList Task list where the tasks from the file are saved.
+     * @throws Elseption If something goes wrong while reading the file.
+     */
+    public void loadFromFile(TaskList taskList) throws Elseption {
```
Good method naming, clearly summarises method

GuanpengR commented on src/main/java/elster/Ui.java:
```diff
@@ -0,0 +1,180 @@
+import elster.tasks.Task;
+
+
+/**
```
Methods are all named appropriately and are clear, nice!

GuanpengR commented on src/main/java/elster/Ui.java:
```diff
@@ -0,0 +1,180 @@
+                + " |    __)_ |  |  /  ___/\\   __\\/ __ \\_  __ \\\n"
+                + " |        \\|  |__\\___ \\  |  | \\  ___/|  | \\/\n"
+                + "/_______  /|____/____  > |__|  \\___  >__|\n"
+                + "        \\/           \\/            \\/";
```
Nice effort with the logo!

SeanFoongjt committed:
Revamps how the Ui class prints error messages

SeanFoongjt committed:
Improve Elseption error messages in Task classes

SeanFoongjt committed:
Implements checkstyle and makes corrections

SeanFoongjt committed:
Merge branch 'branch-A-CheckStyle'

SeanFoongjt committed:
Implement barebones JavaFX features

SeanFoongjt committed:
Update Ui class to return String for message

SeanFoongjt committed:
Implement the logic of Elster into the GUI

SeanFoongjt committed:
Merge branch 'branch-Level-10'

bryanckh commented on src/main/java/elster/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+package elster;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import elster.tasks.Task;
+
```
Methods are all clear, well done!

bryanckh commented on src/main/java/elster/Elster.java:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+
+    /**
+     * Run function of the Elster instance, handles the loading of the save file from storage as
+     * well as general logic of the chatbot.
+     */
```
Javadocs is appropriate and clear, well done!

bryanckh commented on src/main/java/elster/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+     * @return List of tasks that have a description matching user input.
+     */
+    public List<Task> findByDescription(String input) {
+        ArrayList<Task> foundTasks = new ArrayList<>();
```
Use of plural form for a collection of objects, nice!

bryanckh commented on src/main/java/elster/tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,84 @@
+        String[] splitInput = input.split("\\s+");
+        LocalDateTime deadline;
+
+        if (input.strip().equals("deadline")) {
+            throw new Elseption("Elster begs of you to have details for your task");
+
+        } else if (!input.contains("/by")) {
+            throw new Elseption("Elster begs of you to have a yknow, deadline with /by");
+
+        }
+
+        try {
+            String deadlineStr = input.substring(input.indexOf("/by") + 4).strip();
+
+            if (deadlineStr.length() == 10) {
+                deadline = LocalDate.parse(deadlineStr).atTime(23, 59);
+
+            } else {
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+                deadline = LocalDateTime.parse(deadlineStr, formatter);
+            }
+
+        } catch (Exception e) {
+            throw new Elseption("for /by, Elster requires a yyyy-mm-dd or yyyy-mm-dd HH:mm "
+                    + "format please and thanks");
+        }
```
Good handling of exceptions, well done!

bryanckh reviewed (commented):
Overall, your code was easy to read with clear documentation and appropriate naming, well done!

SeanFoongjt committed:
Add Java asserts to Storage class

SeanFoongjt committed:
Add Java assert to TaskList, Ui and Elster

SeanFoongjt committed:
Check asserts created and their messages

There is currently no asserts in the code for Elster.

This means that we do not verify our assumptions about the program state
at runtime and can cause undesirable details to be exposed to the client
from error messages.

Let's add asserts into the code so that we can verify that key
assumptions stay true as well as prevent exposing undesirable details to
our clients.

This can also ensure that regression testing can be easier done when
modifications are made in the future.

SeanFoongjt committed:
Extract methods in Storage class

Storage class have repeated if else structures. Additionally, there
were long methods that had multiple levels of abstraction.

The repeated behaviours caused code duplication, whereas the violation
of SLAP made the code difficult to read for someone new to the codebase.

Let's extracts behaviours common to the methods in the Storage class to
address code duplication. Let's also extract methods such that SLAP is
obeyed by the code.

This ensures that the code is more readable to anyone that comes accross
the codebase.

SeanFoongjt committed:
Extract common behaviour in Elster to Commands

SeanFoongjt committed:
Add JavaDocs comments to Commands

SeanFoongjt committed:
Merge pull request #3 from SeanFoongjt/branch-A-Assertions

Branch A-Assertions

SeanFoongjt committed:
Merge branch 'master' into branch-A-CodeQuality

SeanFoongjt committed:
Merge pull request #2 from SeanFoongjt/branch-A-CodeQuality

Branch A-CodeQuality

SeanFoongjt committed:
Add functionality for tagging a task

SeanFoongjt committed:
Add untag functionality

SeanFoongjt committed:
Merge branch 'branch-BCD-Extension'

SeanFoongjt committed:
Merge branch 'branch-BCD-Extension'

SeanFoongjt committed:
Improve GUI

SeanFoongjt committed:
Change code for better coding quality

SeanFoongjt committed:
Improve code quality

SeanFoongjt committed:
Add user guide

SeanFoongjt committed:
Change colour of README

SeanFoongjt committed:
Add photos for user guide

SeanFoongjt committed:
Correct mistake in user guide

SeanFoongjt committed:
Fix problems related to images in user guide
